File: day11/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	file, _ := os.Open(filename)
	defer file.Close()

	graph := make(map[string][]string)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ": ")
		node := parts[0]
		outputs := strings.Fields(parts[1])
		graph[node] = outputs
	}

	// State: (node, visitedMask) where mask tracks if dac(bit 0) and fft(bit 1) visited
	// memo[node][mask] = number of paths from node to out with that mask
	memo := make(map[string]map[int]int)

	var countPaths func(node string, mask int) int
	countPaths = func(node string, mask int) int {
		// Update mask based on current node
		if node == "dac" {
			mask |= 1
		}
		if node == "fft" {
			mask |= 2
		}

		if node == "out" {
			// Only count if both dac and fft were visited
			if mask == 3 {
				return 1
			}
			return 0
		}

		if memo[node] == nil {
			memo[node] = make(map[int]int)
		}
		if val, exists := memo[node][mask]; exists {
			return val
		}

		total := 0
		for _, next := range graph[node] {
			total += countPaths(next, mask)
		}

		memo[node][mask] = total
		return total
	}

	fmt.Println(countPaths("svr", 0))
}

File: day11/haskell/part2.hs
import System.Environment (getArgs)
import qualified Data.Map as Map
import Data.Bits (setBit, testBit)

parseLine :: String -> (String, [String])
parseLine line = (node, outputs)
  where
    (node, rest) = break (== ':') line
    outputs = words $ drop 2 rest

-- State key: (node, mask) where mask tracks visited required nodes
type MemoKey = (String, Int)
type Memo = Map.Map MemoKey Int

-- Update mask based on node
updateMask :: String -> Int -> Int
updateMask "dac" m = setBit m 0
updateMask "fft" m = setBit m 1
updateMask _ m = m

countPaths :: Map.Map String [String] -> String -> Int -> Memo -> (Int, Memo)
countPaths graph node mask memo =
  let mask' = updateMask node mask
  in if node == "out"
     then (if mask' == 3 then 1 else 0, memo)
     else case Map.lookup (node, mask') memo of
       Just val -> (val, memo)
       Nothing ->
         let outputs = Map.findWithDefault [] node graph
             (total, memo') = foldl addPath (0, memo) outputs
         in (total, Map.insert (node, mask') total memo')
  where
    mask' = updateMask node mask
    addPath (acc, m) next =
      let (count, m') = countPaths graph next mask' m
      in (acc + count, m')

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "input.txt" else head args
  content <- readFile filename
  let graph = Map.fromList $ map parseLine $ lines content
      (result, _) = countPaths graph "svr" 0 Map.empty
  print result
